### DISCUSSION TITLE
Would it be possible to use value tuples for InlineData?

### DISCUSSION THREAD

**Discussion by holmes-mike** [xunit/xunit#2715] [Question] [answered]
I note that array creation is a valid for inlineData. I'm curious as to why ValueTuples are not? I work with geometry a fair bit and being able to do the following would be quite nice.

```csharp
[Theory]
[InlineData((0.0, 0.0, 0.0), (1.0, 0.0, 0.0))]
public void Test((double X, double Y, double Z) p1, (double X, double Y, double Z) p2)
{
...
}
```

**Comment by bradwilson** [comment, marked as answer]
This is a compiler limitation, and has nothing to do with xUnit.net:

`error CS0182: An attribute argument must be a constant expression, typeof expression or array creation expression of an attribute parameter type`

**Comment by mwpowellhtx** [comment]
On a somewhat related note, I experimented with a tuplized set of `Theory` parameters, working around an issue with such argument passing failing to convert from `object` things like `int` in the test case to `int?` in the method signature. So instead I bundled them in a tuple, i.e. `(int? x)`, and it works, sort of. Note that is a very oversimplified bundle; we're talking more like `(string, string, decimal?, int?, bool?, short?)`, do not ask why all those are necessary, they are necessary. However, then we do not get test case discovery. We do see cases appear to be discoverable at some level, and we are getting the iterations run, but we do not see them enumerated in the _Test Explorer_. IIRC, memory a bit fuzzy there, but isn't there a way of telling _xUnit_ how to pretty print arguments under cases such as these, i.e. perhaps emulating method arguments as intended. Cheers, thanks... 🍻

**Reply by mwpowellhtx** [reply]
To clarify, how I got that to work was via `MemberData`, not `InlineData`. But then I ran into discovery failures around the nullable types, leading to me to the tuple experiment.
